New Coding Style guide (#95)
pyansys/dev-guide: doc/source/coding-style/formatting-tools.rst

Code Style Tools
================

> There are plenty of tools for checking code style. This section presents some of
> the most popular ones in the Python ecosystem. A minimum configuration is
> provided for each one so you can easily include them in your PyAnsys project.

Reviewer: ```suggestion
There are many tools for checking code style. This section presents some of
the most popular ones in the Python ecosystem. A minimum configuration is
provided for each one so that you can easily include them in your PyAnsys project.
```

> Most of the tools presented can be configured using :ref:`the
> \`\`pyproject.toml\`\` file`, avoiding dotfiles and thus leading to a much
> cleaner root project directory.

Reviewer: ```suggestion
Most of the tools presented can be configured using :ref:`the
\`\`pyproject.toml\`\` file`. Avoiding dotfiles leads to a much
cleaner root project directory.
```

Black
-----

> `Black`_ is the most popular code formatter in the Python community, as it is
> being maintained by the Python Software Foundation. It allows for a minimum
> configuration to ensure that Python code format looks almost the same across
> projects. 

Reviewer: ```suggestion
`Black`_ is the most popular code formatter in the Python community because it is
maintained by the Python Software Foundation. It allows for a minimum
configuration to ensure that the Python code format looks almost the same across
projects. 
```

> Unlike `PEP 8`_, the default line length imposed by `black`_ is 88 and not 79
> characters.

Reviewer: ```suggestion
While `PEP 8`_ imposes a default line length of 79 characters, `black`_ has
a default line length of  88 characters.
```

> The minimum black configuration for a PyAnsys project should look like:

Reviewer: ```suggestion
The minimum `black`_ configuration for a PyAnsys project should look like this:
```

.. code-block:: toml

    [tool.black]
    line-length: "<length>"

Isort
-----

> The goal of `isort`_  is to properly format ``import`` statements by making sure
> they follow the standard library, third party libraries and custom library
> order.

Reviewer: ```suggestion
The goal of `isort`_  is to properly format ``import`` statements by making sure
that they follow the standard order: library, third-party libraries, and custom library.
```

When using `isort`_ with `black`_, it is important to properly configure both

> tools so no conflicts appear. To do so, make sure you take advantage of the

Reviewer: ```suggestion
tools so that no conflicts appear. To do so, make sure you take advantage of the
```

``--porfile black`` flag in `isort`_.

.. code-block:: toml

   [tool.isort]
   profile = "black"
   force_sort_within_sections = true
   line_length = "<length>"
   default_section = "THIRDPARTY"
   src_paths = ["doc", "src", "tests"]

Flake8
------

> The goal of `flake8` is to act as a `PEP 8`_ compliance checker. Again, it is
> important to make sure that if this tool is being used with `black`_, no
> conflicts arise.

Reviewer: ```suggestion
The goal of `flake8` is to act as a `PEP 8`_ compliance checker. Again, if
this tool is being used with `black`_, it is important to make sure that no
conflicts arise.
```

> The following configuration is the minimum one to setup `flake8`_ together with

Reviewer: ```suggestion
The following configuration is the minimum one to set up `flake8`_ together with
```

> The following configuration is the minimum one to setup `flake8`_ together with
> `black`_ one.

Reviewer: ```suggestion
The following configuration is the minimum one to set up `flake8`_ together with
`black`_.
```

The configuration for `flake8`_ must be specified in a ``.flake8`` file.

.. code-block:: toml

   [flake8]
   max-line-length = 88
   extend-ignore = E203

Pre-commit
----------

> To make sure that every commit you made is compliant with the code style
> guidelines of PyAnsys, you can take advantage of `pre-commit`_ in your project.
> Every time you stage some changes and once you commit those, `pre-commit`_ will
> only allow you to do so if all the defined hooks succeedd.

Reviewer: ```suggestion
To ensure that every commit you make is compliant with the code style
guidelines for PyAnsys, you can take advantage of `pre-commit`_ in your project.
Every time you stage some changes and try to commit them, `pre-commit`_ will
only allow you to do this only if all defined hooks succeed.
```

The configuration for `pre-commit`_ must be defined in a
`.pre-commit-config.yaml` file. The following lines present a minimum

> `pre-commit`_ configuration which includes both code and documentation

Reviewer: ```suggestion
`pre-commit`_ configuration that includes both code and documentation
```

formatting tools.

.. code-block:: yaml

    repos:
    
    - repo: https://github.com/psf/black
      rev: X.Y.Z
      hooks:
      - id: black
    
    - repo: https://github.com/pycqa/isort
      rev: X.Y.Z
      hooks:
      - id: isort
    
    - repo: https://github.com/PyCQA/flake8
      rev: X.Y.Z
      hooks:
      - id: flake8
    
    - repo: https://github.com/codespell-project/codespell
      rev: vX.Y.Z
      hooks:
      - id: codespell
    
    - repo: https://github.com/pycqa/pydocstyle
      rev: X.Y.Z
      hooks:
      - id: pydocstyle
        additional_dependencies: [toml]
        exclude: "tests/"

Installing ``pre-commit``
~~~~~~~~~~~~~~~~~~~~~~~~~
You can install ``pre-commit`` by running:

.. code-block:: bash

    python -m pip install pre-commit

Then, make sure you install it as a ``Git hook`` by running:

.. code-block:: bash

    pre-commit install

Using ``pre-commit``
~~~~~~~~~~~~~~~~~~~~

> From then on, pre-commit will automatically trigger every time you try to commit
> a change. If any of the hooks defined in `.pre-commit-config.yaml` fails, you
> will need to fix the failing files, stage the new changes and try to commit
> those again.

Reviewer: ```suggestion
One installed as described, ``pre-commit`` will automatically trigger every time
that you try to commit a change. If any of the hooks defined in `.pre-commit-config.yaml`
fails, you must fix the failing files, stage the new changes, and try to commit
them again.
```

> If you want to manually run ``pre-commit``, you can execute:

Reviewer: ```suggestion
If you want to manually run ``pre-commit``, you can run:
```

.. code-block:: bash

    pre-commit run --all-files --show-diff-on-failure

> Previous command will show the current and expected style of the code if any of

Reviewer: ```suggestion
This command will show the current and expected style of the code if any of
```

the hooks fail.

> Using ``pre-commit``

Reviewer: This same exact title heading was used earlier. Title headings should be unique.

~~~~~~~~~~~~~~~~~~~~

Tox
---

> A tool you may consider to use in your project is `tox`_. This tool is an
> automation tool similar to `Make`_ but with the advantage of allowing to test
> your package in a temporary virtual environment. This guarantees reproducible
> builds, as your package is no longer tested in "local" mode but in isolated
> form.

Reviewer: ```suggestion
A tool you might consider using in your project is `tox`_. While this automation
tool is similar to `Make`_, it supports testing of your package in a temporary
virtual environment. Being able to test your package in isolation rather than in
"local"  mode guarantees reproducible builds.
```

    skip_missing_interpreters = true
    isolated_build = true
    isolated_build_env = build
    
    [testenv]
    description = Checks for project unit tests and coverage (if desired)
    basepython =
        py37: python3.7
        py38: python3.8
        py39: python3.9
        py310: python3.10
        py: python3
        {style,reformat,doc,build}: python3
    setenv =
        PYTHONUNBUFFERED = yes
        coverage: PYTEST_EXTRA_ARGS = --cov=ansys.product --cov-report=term --cov-report=xml --cov-report=html
    deps =
        -r{toxinidir}/requirements/requirements_tests.txt
    commands =
        pytest {env:PYTEST_MARKERS:} {env:PYTEST_EXTRA_ARGS:} {posargs:-vv}
    
    [testenv:style]
    description = Checks project code style
    skip_install = true
    deps =
        pre-commit
    commands =
        pre-commit install
        pre-commit run --all-files --show-diff-on-failure
    
    [testenv:doc]
    description = Check if documentation generates properly
    deps =
        -r{toxinidir}/requirements/requirements_doc.txt
    commands =
        sphinx-build -d "{toxworkdir}/doc_doctree" doc/source "{toxworkdir}/doc_out" --color -vW -bhtml

> Previous configuration assumes that you have a ``requirements/`` directory that
> contains a ``requirements_tests.txt`` and a ``requirements_doc.txt``. In
> addition, the ``style`` environment will execute pre-commit, which guarantees
> the usage of this tool in your project.

Reviewer: ```suggestion
This minimum configuration assumes that you have a ``requirements/`` directory that
contains ``requirements_tests.txt`` and ``requirements_doc.txt``. In
addition, the ``style`` environment will execute ``pre-commit``, which guarantees
the usage of this tool in your project.
```

Installing ``tox``
~~~~~~~~~~~~~~~~~~

> You can install this tool as any other Python one by running:

Reviewer: ```suggestion
You can install ``tox`` like any other Python package:
```

.. code-block:: bash

    python -m pip install tox

Using ``tox``
~~~~~~~~~~~~~

> The core concept behind `tox`_ are ``environments``. These are similar to
> ``Makefile`` rules and highly customizable. Previous configuration ships with
> different environments among which you can find:

Reviewer: ```suggestion
Behind `tox`_ are ``environments``, which are similar to
``Makefile`` rules and highly customizable. The previous configuration ships with
different environments, including:
```

Reviewer: I wasn't sure why "previous" was present in "The previous configuration ...." This sentence might need to be clarified.

- ``style``: for checking the code style of your project.

> - ``py``: which will run your test suite.

Reviewer: ```suggestion
- ``py``: for running your test suite.
```

- ``doc``: for building the documentation of your project.

Execute any of the previous environments by running ``tox -e <env-name>``. You

> can run multiple environments by specifying those with commas ``tox -e

Reviewer: ```suggestion
can run multiple environments by separating them with commas ``tox -e
```

<env-name0>,<env-name1>,...```.  To run all available environments, simply

> execute ``tox``.

Reviewer: ```suggestion
run ``tox``.
```

Code Style Tools
================

There are many tools for checking code style. This section presents some of
the most popular ones in the Python ecosystem. A minimum configuration is
provided for each one so that you can easily include them in your PyAnsys project.

Most of the tools presented can be configured using :ref:`the
\`\`pyproject.toml\`\` file`. Avoiding dotfiles leads to a much
cleaner root project directory.

Black
-----
`Black`_ is the most popular code formatter in the Python community because it is
maintained by the Python Software Foundation. It allows for a minimum
configuration to ensure that the Python code format looks almost the same across
projects. 

While `PEP 8`_ imposes a default line length of 79 characters, `black`_ has
a default line length of  88 characters.

The minimum `black`_ configuration for a PyAnsys project should look like this:

.. code-block:: toml

    [tool.black]
    line-length: "<length>"

Isort
-----
The goal of `isort`_  is to properly format ``import`` statements by making sure

> that they follow the standard order: library, third-party libraries, and custom library.

Reviewer: ```suggestion
that they follow the standard order: library, third-party libraries, and custom libraries.
```

When using `isort`_ with `black`_, it is important to properly configure both

> tools so that no conflicts appear. To do so, make sure you take advantage of the

Reviewer: ```suggestion
tools so that no conflicts occur. To accomplish this, make sure you take advantage of the
```

``--porfile black`` flag in `isort`_.

.. code-block:: toml

   [tool.isort]
   profile = "black"
   force_sort_within_sections = true
   line_length = "<length>"
   default_section = "THIRDPARTY"
   src_paths = ["doc", "src", "tests"]

Flake8
------
The goal of `flake8` is to act as a `PEP 8`_ compliance checker. Again, if
this tool is being used with `black`_, it is important to make sure that no
conflicts arise.

The following configuration is the minimum one to set up `flake8`_ together with
`black`_.

The configuration for `flake8`_ must be specified in a ``.flake8`` file.

.. code-block:: toml

   [flake8]
   max-line-length = 88
   extend-ignore = E203

Flake8 has many options that can be set within the configuration file.

> For a list and descriptions, see this `Flake8 documentation topic

Reviewer: ```suggestion
For more information, see this `Flake8 documentation topic
```

<https://flake8.pycqa.org/en/latest/user/options.html>`__.

> The example configuration defines the following options:

Reviewer: ```suggestion
The example configuration defines these options:
```

- ``exclude``

>     Denotes subdirectories and files to be excluded from the check.

Reviewer: ```suggestion
    Subdirectories and files to exclude from the check.
```

- ``select``

>     Sequence of error codes that Flake8 will report errors

Reviewer: ```suggestion
    Sequence of error codes that Flake8 is to report errors
```

    for. The set in the above configuration is a basic set of errors to
    check for and is not an exhaustive list.

    For a full list of error codes and their descriptions, see this `Flake8
    documentation topic <https://flake8.pycqa.org/en/3.9.2/user/error-codes.html>`__.

- ``count``
    Total number of errors to print at the end of the check.

- ``max-complexity``

>     Sets the maximum allowed McCabe complexity value for a block of code.

Reviewer: ```suggestion
   Maximum allowed McCabe complexity value for a block of code.
```

    The value of 10 was chosen because it is a common default.

- ``statistics``
    Number of occurrences of each error or warning code

>     to be printed as a report at the end of the check.

Reviewer: ```suggestion
    to print as a report at the end of the check.
```

Code Coverage
-------------

> Code coverage allows to check the percetange of codebase tested by the test
> suite. Code coverage should be as high as possible to guarantee that every piece
> of code has been tested.

Reviewer: ```suggestion
Code coverage indicates the percentage of the codebase tested by the test
suite. Code coverage should be as high as possible to guarantee that every piece
of code has been tested.
```

For ``PyAnsys``, code coverage is done using `pytest-cov`_, a `pytest`_ plugin

> which will trigger the code coverage analysis once your test suite has executed.

Reviewer: ```suggestion
that triggers the code coverage analysis once your test suite has executed.
```

Considering the layout presented in :ref:`Required Files`, the following
configuration for code coverage is the minimum one required for a ``PyAnsys``
project:

.. code-block:: toml

   [tool.coverage.run]
   source = ["ansys.<product>"]

   [tool.coverage.report]
   show_missing = true

Pre-commit
----------
To ensure that every commit you make is compliant with the code style
guidelines for PyAnsys, you can take advantage of `pre-commit`_ in your project.

> Every time you stage some changes and try to commit them, `pre-commit`_ will
> only allow you to do this only if all defined hooks succeed.

Reviewer: ```suggestion
Every time you stage some changes and try to commit them, `pre-commit`_ only
allows them to be committed if all defined hooks succeed.
```

The configuration for `pre-commit`_ must be defined in a
``.pre-commit-config.yaml`` file. The following lines present a minimum
`pre-commit`_ configuration that includes both code and documentation
formatting tools.

.. code-block:: yaml

    repos:
    
    - repo: https://github.com/psf/black
      rev: X.Y.Z
      hooks:
      - id: black
    
    - repo: https://github.com/pycqa/isort
      rev: X.Y.Z
      hooks:
      - id: isort
    
    - repo: https://github.com/PyCQA/flake8
      rev: X.Y.Z
      hooks:
      - id: flake8
    
    - repo: https://github.com/codespell-project/codespell
      rev: vX.Y.Z
      hooks:
      - id: codespell
    
    - repo: https://github.com/pycqa/pydocstyle
      rev: X.Y.Z
      hooks:
      - id: pydocstyle
        additional_dependencies: [toml]
        exclude: "tests/"

Installing ``pre-commit``
~~~~~~~~~~~~~~~~~~~~~~~~~
You can install ``pre-commit`` by running:

.. code-block:: bash

    python -m pip install pre-commit

> Then, make sure you install it as a ``Git hook`` by running:

Reviewer: ```suggestion
Then, ensure that you install it as a ``Git hook`` by running:
```

.. code-block:: bash

    pre-commit install

Using ``pre-commit``
~~~~~~~~~~~~~~~~~~~~

> One installed as described, ``pre-commit`` will automatically trigger every time

Reviewer: ```suggestion
One installed as described, ``pre-commit`` automatically triggers every time
```

> that you try to commit a change. If any of the hooks defined in `.pre-commit-config.yaml`

Reviewer: ```suggestion
that you try to commit a change. If any hook defined in `.pre-commit-config.yaml`
```

fails, you must fix the failing files, stage the new changes, and try to commit
them again.

If you want to manually run ``pre-commit``, you can run:

.. code-block:: bash

    pre-commit run --all-files --show-diff-on-failure

This command will show the current and expected style of the code if any of
the hooks fail.

Tox
---

> A tool you might consider using in your project is `tox`_. While this automation

Reviewer: ```suggestion
You might consider using `tox`_ in your project. While this automation
```

tool is similar to `Make`_, it supports testing of your package in a temporary
virtual environment. Being able to test your package in isolation rather than in

> "local"  mode guarantees reproducible builds.

Reviewer: ```suggestion
"local" mode guarantees reproducible builds.
```

Configuration for `tox`_ is stored in a ``tox.ini`` file. The minimum
configuration for a PyAnsys ``py<product>-<library>`` project should be:

.. tabs::

    .. tab:: Tox with Flit

        .. include:: code/tox-flit.rst

    .. tab:: Tox with Poetry

        .. include:: code/tox-poetry.rst

This minimum configuration assumes that you have a ``requirements/`` directory that
contains ``requirements_tests.txt`` and ``requirements_doc.txt``. In

> addition, the ``style`` environment will execute ``pre-commit``, which guarantees

Reviewer: ```suggestion
addition, the ``style`` environment must execute ``pre-commit``, which guarantees
```
